src/main/java/org/opensearch/knn/index/util/Faiss.java

```diff
@@ -226,7 +228,28 @@ class Faiss extends NativeLibrary {
                     return ((4L * centroids * dimension) / BYTES_PER_KILOBYTES) + 1;
                 })
                 .build()
-        ).addSpaces(SpaceType.L2, SpaceType.INNER_PRODUCT).build()
+        ).addSpaces(SpaceType.L2, SpaceType.INNER_PRODUCT).build(),
+        METHOD_NSG,
+        KNNMethod.Builder.builder(
+            MethodComponent.Builder.builder(METHOD_NSG)
+                .addParameter(
+                    METHOD_PARAMETER_M,
+                    new Parameter.IntegerParameter(METHOD_PARAMETER_M, KNNSettings.INDEX_KNN_DEFAULT_ALGO_PARAM_M, v -> v > 0)
+                )
+                .addParameter(
```

Comment:
All existing encoders work?

Reply:
@luyuncheng Can you confirm this?

Reply:
In https://github.com/facebookresearch/faiss/blob/3219e3d12e6fc36dfdfe17d4cf238ef70bf89568/faiss/index_factory.cpp#L493 it can support PQ parameter, so i think encoder parameter is ok. i also added unit tests in https://github.com/opensearch-project/k-NN/pull/945/commits/513f38cb3d38ef0fd1a2762866d911585d7d76e8

```
    if (match("PQ([0-9]+)(np)?")) {
        int M = std::stoi(sm[1].str());
        IndexNSGPQ* ipq = new IndexNSGPQ(d, M, nsg_R);
        dynamic_cast<IndexPQ*>(ipq->storage)->do_polysemous_training =
                sm[2].str() != "np";
        return ipq;
    }
```